diff -up udev-147/extras/modem-modeswitch/61-option-modem-modeswitch.rules.orig udev-147/extras/modem-modeswitch/61-option-modem-modeswitch.rules
--- udev-147/extras/modem-modeswitch/61-option-modem-modeswitch.rules.orig	2009-07-31 12:48:49.000000000 +0200
+++ udev-147/extras/modem-modeswitch/61-option-modem-modeswitch.rules	2009-11-12 10:47:49.000000000 +0100
@@ -35,5 +35,6 @@ ATTRS{idVendor}=="0af0", ATTRS{idProduct
 ATTRS{idVendor}=="0af0", ATTRS{idProduct}=="7601", RUN+="modem-modeswitch -v 0x%s{idVendor} -p 0x%s{idProduct} -t option-zerocd"
 ATTRS{idVendor}=="0af0", ATTRS{idProduct}=="7901", RUN+="modem-modeswitch -v 0x%s{idVendor} -p 0x%s{idProduct} -t option-zerocd"
 ATTRS{idVendor}=="12d1", ATTRS{idProduct}=="1446", RUN+="modem-modeswitch -v 0x%s{idVendor} -p 0x%s{idProduct} -t option-zerocd"
+ATTRS{idVendor}=="19d2", ATTRS{idProduct}=="2000", RUN+="modem-modeswitch -v 0x%s{idVendor} -p 0x%s{idProduct} -t option-zerocd"
 
 LABEL="option_zerocd_end"
