diff -up udev-147/extras/cdrom_id/cdrom_id.c.virt udev-147/extras/cdrom_id/cdrom_id.c
--- udev-147/extras/cdrom_id/cdrom_id.c.virt	2010-08-24 16:50:16.000000000 +0200
+++ udev-147/extras/cdrom_id/cdrom_id.c	2010-08-31 16:12:38.000000000 +0200
@@ -359,7 +359,15 @@ static int cd_profiles(struct udev *udev
 	/* current media profile */
 	cur_profile = header[6] << 8 | header[7];
 	info(udev, "current profile 0x%02x\n", cur_profile);
-	if (cur_profile == 0) {
+	if (cur_profile == 0) {	   
+		if (cd_media == 1) {
+			/* quirk for xen virtual machines */
+			info(udev, "no current profile, but CDS_DISC_OK, assuming incomplete implementation\n");
+			cd_media_track_count_data = 1;
+			cd_media_track_count = 1;
+			cd_media_cd_rom = 1;
+			return 0;
+		}
 		info(udev, "no current profile, assuming no media\n");
 		return -1;
 	}
@@ -634,15 +642,15 @@ int main(int argc, char *argv[])
 		goto print;
 
 	/* read drive and possibly current profile */
-	if (cd_profiles(udev, fd) < 0)
+	if (cd_profiles(udev, fd) < 0 && cd_media != 1)
 		goto print;
 
 	/* get session/track info */
-	if (cd_media_toc(udev, fd) < 0)
+	if (cd_media_toc(udev, fd) < 0 && cd_media != 1)
 		goto print;
 
 	/* get writable media state */
-	if (cd_media_info(udev, fd) < 0)
+	if (cd_media_info(udev, fd) < 0 && cd_media != 1)
 		goto print;
 
 print:
