diff -up udev-147/rules/rules.d/60-persistent-storage.rules.orig udev-147/rules/rules.d/60-persistent-storage.rules
--- udev-147/rules/rules.d/60-persistent-storage.rules.orig	2010-06-29 12:47:27.000000000 +0200
+++ udev-147/rules/rules.d/60-persistent-storage.rules	2010-06-29 12:47:46.000000000 +0200
@@ -27,6 +27,10 @@ KERNEL=="hd*[!0-9]", IMPORT{program}="at
 KERNEL=="hd*[!0-9]", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/ata-$env{ID_SERIAL}"
 KERNEL=="hd*[0-9]", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/ata-$env{ID_SERIAL}-part%n"
 
+# virtio-blk
+KERNEL=="vd*[!0-9]", ATTRS{serial}=="?*", ENV{ID_SERIAL}="$attr{serial}", SYMLINK+="disk/by-id/virtio-$env{ID_SERIAL}"
+KERNEL=="vd*[0-9]", ATTRS{serial}=="?*", ENV{ID_SERIAL}="$attr{serial}", SYMLINK+="disk/by-id/virtio-$env{ID_SERIAL}-part%n"
+
 # USB devices use their own serial number
 KERNEL=="sd*[!0-9]|sr*", ENV{ID_SERIAL}!="?*", SUBSYSTEMS=="usb", IMPORT{program}="usb_id --export %p"
 # ATA devices with their own "ata" kernel subsystem
